src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/MethodBodyStreamEncoder.cs

```diff
@@ -0,0 +1,204 @@
+            {
+                Throw.BuilderArgumentNull();
+            }
+
+            // Fat methods are 4-byte aligned. We calculate the alignment relative to the start of the ILStream.
+            //
+            // See ECMA-335 paragraph 25.4.5, Method data section:
+            // "At the next 4-byte boundary following the method body can be extra method data sections."
+            if ((builder.Count % 4) != 0)
+            {
+                throw new ArgumentException(SR.BuilderMustAligned, nameof(builder));
+            }
+
+            Builder = builder;
+        }
+
+        /// <summary>
+        /// Encodes a method body and adds it to the method body stream.
+        /// </summary>
+        /// <param name="codeSize">Number of bytes to be reserved for instructions.</param>
+        /// <param name="maxStack">Max stack.</param>
+        /// <param name="exceptionRegionCount">Number of exception regions.</param>
+        /// <param name="hasFatExceptionRegions">True if the exception regions should be encoded in 'fat' format.</param>
+        /// <param name="localVariablesSignature">Local variables signature handle.</param>
+        /// <param name="attributes">Attributes.</param>
+        /// <returns>The offset of the encoded body within the method body stream.</returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="codeSize"/>, <paramref name="exceptionRegionCount"/>, or <paramref name="maxStack"/> is out of allowed range.
+        /// </exception>
+        public MethodBody AddMethodBody(
+            int codeSize,
+            int maxStack = 8,
+            int exceptionRegionCount = 0,
+            bool hasFatExceptionRegions = false,
+            StandaloneSignatureHandle localVariablesSignature = default(StandaloneSignatureHandle),
+            MethodBodyAttributes attributes = MethodBodyAttributes.InitLocals)
+        {
+            if (codeSize < 0)
+            {
+                Throw.ArgumentOutOfRange(nameof(codeSize));
+            }
+
+            if (unchecked((uint)maxStack) > ushort.MaxValue)
+            {
+                Throw.ArgumentOutOfRange(nameof(maxStack));
+            }
+
+            if (!ExceptionRegionEncoder.IsExceptionRegionCountInBounds(exceptionRegionCount))
+            {
+                Throw.ArgumentOutOfRange(nameof(exceptionRegionCount));
+            }
+
+            int bodyOffset = SerializeHeader(codeSize, (ushort)maxStack, exceptionRegionCount, attributes, localVariablesSignature);
+            var instructions = Builder.ReserveBytes(codeSize);
+
+            var regionEncoder = (exceptionRegionCount > 0) ? 
+                ExceptionRegionEncoder.SerializeTableHeader(Builder, exceptionRegionCount, hasFatExceptionRegions) : default(ExceptionRegionEncoder);
+
+            // return a tuple
+            return new MethodBody(bodyOffset, instructions, regionEncoder);
+        }
+
+        public struct MethodBody
```

Comment:
That's the problem with not having ValueTuples - I don't want to pollute the namespace with this struct. I don't think anybody will need to actually use the type. Typically the usage would be either ```var (offset, instr, ex) = AddMethodBody()``` in C# 7 or ```var body = AddMethodBody()```. I'd also leave it to the API review though.